src/fighter_state.rs

```diff
@@ -0,0 +1,747 @@
+        // Transition to any higher priority states
+        let current_state_removed = transition_intents
+            .transition_to_higher_priority_states::<Attacking>(
+                entity,
+                Attacking::PRIORITY,
+                &mut commands,
+            );
+
+        // If our current state was removed, don't continue processing this fighter
+        if current_state_removed {
+            continue 'entity;
+        }
+
+        // If we're done flopping
+        if flopping.is_finished {
+            // Go back to idle
+            commands.entity(entity).remove::<Attacking>().insert(Idling);
+        }
+    }
+}
+
+// Initiate any transitions from the knocked back state
+fn transition_from_knocked_back(
+    mut commands: Commands,
+    mut fighters: Query<(Entity, &mut StateTransitionIntents, &KnockedBack)>,
+) {
+    'entity: for (entity, mut transition_intents, knocked_back) in &mut fighters {
+        // Transition to any higher priority states
+        let current_state_removed = transition_intents
+            .transition_to_higher_priority_states::<KnockedBack>(
+                entity,
+                KnockedBack::PRIORITY,
+                &mut commands,
+            );
+
+        // If our current state was removed, don't continue processing this fighter
+        if current_state_removed {
+            continue 'entity;
+        }
+
+        // Transition to idle when finished
+        if knocked_back.timer.finished() {
+            commands
+                .entity(entity)
+                .remove::<KnockedBack>()
+                .insert(Idling);
+        }
+    }
+}
+
+//
+// Handle state systems
+//
+
+/// Handle fighter idle state
+fn idling(mut fighters: Query<(&mut Animation, &mut LinearVelocity), With<Idling>>) {
+    for (mut animation, mut velocity) in &mut fighters {
+        // If we aren't playing the idle animation
+        if animation.current_animation.as_deref() != Some(Idling::ANIMATION) {
+            // Start the idle animation from the beginning
+            animation.play(Idling::ANIMATION, true /* repeating */)
+        }
+
+        // Stop moving playe when we idle
+        **velocity = Vec2::ZERO;
+    }
+}
+
+/// Handle fighter attacking state
+///
+/// > **Note:** This system currently applies attacks for both enemies and players, doing a sort of
+/// > jumping "punch". In the future there will be different attacks, which will each have their own
+/// > state system, and we will trigger different attack states for different players and enemies,
+/// > based on the attacks available to that fighter.
+fn attacking(
+    mut commands: Commands,
+    mut fighters: Query<(
+        Entity,
+        &mut Animation,
+        &mut Transform,
+        &mut LinearVelocity,
+        &Facing,
+        &Stats,
+        &Handle<FighterMeta>,
+        &mut Attacking,
+        Option<&Player>,
+        Option<&Enemy>,
```

Comment:
Could potentially use an Or filter

Reply:
I don't think I could tell which one is present if we used an or filter, though. I need to call `player.is_some()` down below.

Reply:
Ohh you are right I didn't actually notice that `is_player, is_enemy` variables were used again after the continue check.